rebuild/src/signature.h

```diff
@@ -0,0 +1,119 @@
+#ifndef BLST_TS_SIGNATURE_H__
+#define BLST_TS_SIGNATURE_H__
+
+#include <memory>
+
+#include "addon.h"
+#include "blst.hpp"
+#include "napi.h"
+
+#define BLST_TS_SIGNATURE_LENGTH_COMPRESSED 96U
+#define BLST_TS_SIGNATURE_LENGTH_UNCOMPRESSED 192U
+
+typedef struct {
+    std::unique_ptr<blst::P2_Affine> smart_pointer;
+    blst::P2_Affine *raw_point;
+} P2AffineGroup;
+
+class P2Wrapper {
+   public:
+    virtual ~P2Wrapper() = default;
+    virtual bool IsInfinite() = 0;
+    virtual bool InGroup() = 0;
+    virtual void Serialize(bool compress, blst::byte *out) = 0;
+    virtual void AddTo(blst::P2 &point) = 0;
+    virtual blst::P2 MultiplyBy(
+        blst::byte *rand_bytes, size_t rand_bytes_length) = 0;
+    virtual P2AffineGroup AsAffine() = 0;
+    virtual void Sign(
+        const blst::SecretKey &key,
```

Comment:
I'd add `const` and `noexcept` where applicable. Potentially `[[nodiscard]]` - this might help avoid errors in the future.

Reply:
I added `const` where I could here: d4122871789720b07446363fd4fec013145a9b32

None of the underlying methods in blst are labeled as noexcept but i was not able to find any exceptions being thrown, however there is a bunch of assembly code that gets called and I am not confident that there are absolutely no exceptions and if on occurs it will be a stop-the-world kind of thing so I don't feel comfortable to add that here.

`[[nodiscard]]` is a C++17 thing and I cannot guarantee that everyone will be using that version or newer to build (although they probably should be).  But I do not want to limit it to that by using this so I am not comfortable adding it in this context.

Reply:
turns out c++17 is necessary for the std::optional you pointed out and I think that feature is worth enforcing 17 for.  I updated the `binding.gyp` so tooling uses the correct compiler version.  I will reinvestigate nodiscard now.

Reply:
Anyone using a pre c++17 compiler should recall what happened to the dinosaurs...

Reply:
Remember, these are not c/c++ devs... they are nodejs end users that often have no idea compilation is even happening.... they are just running `npm install` and "magic happens" to make there code run.  While i would guess that most base docker images and VM instances have c++17 it is not guaranteed but what I will guess is we will get a github issue or discord message "that it doesnt install" and the error should be relatively easy to spot.  backing out the changes if there are many complaints or getting it installed (hopefully will only be 1-off) maybe not so much but will cross that bridge if we get there

Reply:
Yeah, fair enough. You could make the same claim for C++14 or even C++11, though. Sadly, quite a bit of systems still sit on old compilers that are difficult to update at this point.

IMO, the important thing is to make an informed decision, with some baseline. For example, Ubuntu LTS spans 5 years. We could make sure that it runs on _at least_ Ubuntu 18.04. Or, given that it's already past the support, make sure that it runs on _at least_ Ubuntu 20.04.

Reply:
I went through and looked but didn't see anyplaces where `[[nodiscard]]` might be helpful.